### PR TITLE
fix(scap): do not access /proc in generic platform code

### PR DESCRIPTION
**Pull request by gnosek**
**What type of PR is this?**

This is a hard question :) Mostly "iterating over the same thing over and over until everybody is happy"

> Uncomment one (or more) `/kind <>` lines:

/kind bug

> /kind cleanup

> /kind design

> /kind documentation

> /kind failing-test

> /kind feature

**Any specific area of the project related to this PR?**

> Uncomment one (or more) `/area <>` lines:

> /area API-version

> /area build

> /area CI

> /area driver-kmod

> /area driver-bpf

> /area driver-modern-bpf

> /area libscap-engine-bpf

> /area libscap-engine-gvisor

> /area libscap-engine-kmod

> /area libscap-engine-modern-bpf

> /area libscap-engine-nodriver

> /area libscap-engine-noop

> /area libscap-engine-source-plugin

> /area libscap-engine-savefile

> /area libscap-engine-udig

/area libscap

> /area libpman

> /area libsinsp

> /area tests

> /area proposals

**Does this PR require a change in the driver versions?**

> /version driver-API-version-major

> /version driver-API-version-minor

> /version driver-API-version-patch

> /version driver-SCHEMA-version-major

> /version driver-SCHEMA-version-minor

> /version driver-SCHEMA-version-patch

**What this PR does / why we need it**:

We want to be able to run source plugins without mounting /proc from the host. Unfortunately, the generic platform code does some /proc accesses, so we need the mount even for source plugins (which do not by themselves require any sort of OS access).

This PR makes the generic platform fully generic (and at the same time, fully no-op). This means we can remove the machine_info code for non-Linux platforms (since they do not have any "real" platform they could use).

This means that if you *do* need machine info, you need to use some other platform rather than generic (presumably the linux one). This can be enabled by passing SCAP_MODE_LIVE as the mode in sinsp::open_plugin and sinsp::open_test_input

**Which issue(s) this PR fixes**:

Fixes #

**Special notes for your reviewer**:

This is definitely not the first time around the block with this, so I'm waiting for the inevitable "revert this" PR ;)

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

### CODE REVIEW

**Review by Andreagit97 — COMMENTED** [review]
Ok IMO the proposed changes are ok! Thank you!

Just a minor note:

```cpp
void sinsp::init()
{
	//
	// Retrieve machine information
	//
	m_machine_info = scap_get_machine_info(m_h);
	if(m_machine_info != NULL)
	{
		m_num_cpus = m_machine_info->num_cpus;
	}
	else
	{
		ASSERT(false);
		m_num_cpus = 0;
	}
       ...
```

When we will use plugins with the `generic_platform`,  in sinsp `m_machine_info` and `m_agent_info`  will be pointers to empty structs so we cannot trust them (the good news is that there is no risk of segfault since in the scap platform we have structs and not pointers :)). So in this fragment of code `m_num_cpus = m_machine_info->num_cpus;` will be probably set to `0` in the plugin case.

**Comment by gnosek** [comment]
We shouldn't crash since the info structs don't contain any pointers. Whether the defaults are OK, Cthulhu knows (libs don't really use these directly AFAIK)

**Comment by Andreagit97** [comment]
> We shouldn't crash since the info structs don't contain any pointers. Whether the defaults are OK, Cthulhu knows (libs don't really use these directly AFAIK)

In sinsp they are used in some parsers like the clone one

```cpp
/* Get pid namespace start ts - convert monotonic time in ns to epoch ts */
child_tinfo->m_pidns_init_start_ts = m_inspector->m_machine_info->boot_ts_epoch;
```

but right now we don't care it since we are in a source plugin scenario without the syscall source enabled :+1:

File: userspace/libscap/scap_platform.c [hunk 1]
```diff
@@ -21,18 +21,10 @@ limitations under the License.
 
 #include "scap.h"
 #include "scap-int.h"
-#include "scap_os_machine_info.h"
 
 int32_t scap_generic_init_platform(struct scap_platform* platform, char* lasterr, struct scap_open_args* oargs)
 {
 	memset(&platform->m_machine_info, 0, sizeof(platform->m_machine_info));
```
[the remark this answers is not in the document]
Comment (Andreagit97) [reply userspace/libscap/scap_platform.c:27]: Just pushed it :)

**Review by Andreagit97 — APPROVED** [review]
/approve

**Review by FedeDP — APPROVED** [review]
/approve

**Comment by incertum** [comment]
Oh hey @gnosek

> Special notes for your reviewer:
> This is definitely not the first time around the block with this, so I'm waiting for the inevitable "revert this" PR ;)

... the day has come ;)

Instead of reverting, would the following be acceptable to get out of the regression for the host metrics https://github.com/falcosecurity/falco/issues/2821 when you run Falco w/ plugin source only, but still on Linux.

- Add a new `sinsp_mode_t` called `SINSP_MODE_PLUGIN_HOST_METRICS` (or similar) that complements `SINSP_MODE_PLUGIN` when running Falco w/ a plugin source only (no syscalls), but also with the Falco metrics framework enabled
- When running in mode `SINSP_MODE_PLUGIN_HOST_METRICS`, allocate a new platform that could be called `scap_linux_host_metrics_alloc_platform` or similar. It would basically be a Linux light platform that would only alloc what is needed in order to have `m_machine_info`, `m_agent_info` and the `iflist` (for a new metrics field in the making) available to the metrics framework.
- Other ideas?

**Comment by gnosek** [comment]
Moved the discussion back to https://github.com/falcosecurity/falco/issues/2821, working on a PR.

**Pull request merged**